### PR TITLE
Fix bio callback backward compatibility

### PR DESCRIPTION
**Pull request by bernd-edlinger**
This is an alternative fix for #5488 

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by sebastianas** [comment]
So if this fixes the issue that I reported/attempted to fix in #5488 then I am totally fine with it.

Do you think that it would make sense to extend the testcase to
  RSA_print(bio, EVP_PKEY_get1_RSA(key), 0);
or do you think that it won't cover more ground/code than it already does?

**Comment by bernd-edlinger** [comment]
Yes, it fixes your issue, and allows BIO_puts to return the correct value.

Additionally it avoids passing pointers to undefined values for BIO_CB_READ
and BIO_CB_WRITE for which *now* is about the last time we can clean up that kind
of error before 1.1.1 get released.

> Do you think that it would make sense to extend the testcase to
RSA_print(bio, EVP_PKEY_get1_RSA(key), 0);

No I think RSA_print uses the same functions, BIO_write and BIO_puts.

But it might be useful to add test cases for BIO_read and BIO_gets,
and various BIO_ctrl code, and for using the new callback as well.
But I think that can be done when an issue is raised with these functions.

File: crypto/bio/bio_lib.c [hunk 1]
```diff
@@ -51,17 +50,17 @@ static long bio_call_callback(BIO *b, int oper, const char *argp, size_t len,
             return -1;
 
         argi = (int)len;
+    }
```
Comment (paulidale) [inline crypto/bio/bio_lib.c]: The comment about ten lines up mentions _do nasty casts_ which the code isn't.  Worth a clean.

**Comment by bernd-edlinger** [comment]
Ping, @openssl/omc 
could one of you apprive this for the upcoming code-freeze?

**Comment by bernd-edlinger** [comment]
Merged, thanks!

**Pull request closed without merging**